lxd/storage/drivers/driver_pure_util.go

```diff
@@ -517,6 +572,475 @@ func (p *pureClient) disconnectHostFromVolume(poolName string, volName string, h
+				err = d.client().updateHost(hostname, []string{iqn})
+				if err != nil {
+					return "", nil, err
+				}
+			} else {
+				revert.Add(func() { _ = d.client().deleteHost(hostname) })
+			}
+		} else {
+			// Hostname already exists with the given IQN.
+			hostname = host.Name
+		}
+	}
+
+	cleanup = revert.Clone().Fail
+	revert.Success()
+	return hostname, cleanup, nil
+}
+
+// connect connects this host with the PureStorge array. Note that the connection can only
+// be established when at least one volume is mapped with the corresponding PureStorage host.
+// The operation is idempotent and returns nil if already connected to the subsystem.
+func (d *pure) connect() (revert.Hook, error) {
+	revert := revert.New()
+	defer revert.Fail()
+
+	// Find the array's qualified name for the configured mode.
+	target, err := d.client().getTarget()
+	if err != nil {
+		return nil, err
+	}
+
+	switch d.config["pure.mode"] {
+	case pureModeISCSI:
+		targetIQN := *target.IQN
+		targetAddr := d.config["pure.array.address"]
+
+		// Try to find an existing iSCSI session.
+		sessionID, err := iscsiFindSession(targetIQN)
+		if err != nil {
+			return nil, err
+		}
+
+		if sessionID != "" {
+			// Already connected to the PureStorage array via iSCSI.
+			// Rescan the session to ensure new volumes are detected.
+			_, err := shared.RunCommand("iscsiadm", "--mode", "session", "--sid", sessionID, "--rescan")
+			if err != nil {
+				return nil, err
+			}
+
+			cleanup := revert.Clone().Fail
+			revert.Success()
+			return cleanup, nil
+		}
+
+		// Attempt to login into discovered iSCSI targets.
+		_, stderr, err := shared.RunCommandSplit(d.state.ShutdownCtx, nil, nil, "iscsiadm", "--mode", "node", "--targetname", targetIQN, "--portal", targetAddr, "--login")
+		if err != nil {
+			return nil, fmt.Errorf("Failed to connect to PureStorage array %q via iSCSI: %s\n%v", targetAddr, stderr, err)
+		}
+	}
+
+	revert.Add(func() { _ = d.disconnect(*target) })
+
+	cleanup := revert.Clone().Fail
+	revert.Success()
+	return cleanup, nil
+}
+
+// disconnect disconnects this host from the given target with a given qualified name.
+func (d *pure) disconnect(target pureTarget) error {
+	switch d.config["pure.mode"] {
+	case pureModeISCSI:
+		targetIQN := *target.IQN
+
+		sessionID, err := iscsiFindSession(targetIQN)
+		if err != nil {
+			return err
+		}
+
+		if sessionID != "" {
+			// Disconnect from the iSCSI target.
+			_, err := shared.RunCommand("iscsiadm", "--mode", "node", "--targetname", targetIQN, "--logout")
+			if err != nil {
+				return fmt.Errorf("Failed disconnecting from PureStorage iSCSI target %q: %w", targetIQN, err)
+			}
+		}
+	}
+
+	return nil
+}
+
+// mapVolume maps the given volume onto this host.
+func (d *pure) mapVolume(vol Volume) error {
+	revert := revert.New()
+	defer revert.Fail()
+
+	volName, err := d.getVolumeName(vol)
+	if err != nil {
+		return err
+	}
+
+	unlock, err := locking.Lock(d.state.ShutdownCtx, d.config["pure.mode"])
```

Comment:
As this appears to be copying following the powerflex driver's lead of serialising access to the storage mapping subsystem I think we need to move this out to a separate set of functions that can be called from both drivers to clearly show that this lock is spanning multiple drivers.

Also, and this is my fault as I clearly missed this on the PF review, but we should add a prefix to the lock name, something like "storage_".

As each implementation is different after the lock, suggest something like we've done before in storage with `MountTask()` taking a function as an argument to avoid forgetting to unlock.